aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -526,41 +526,101 @@ static Tensor& linalg_solve_out_info(Tensor& result, Tensor& infos, const Tensor
   TORCH_CHECK(other.dim() >= 1,
            "other should have at least 1 dimension, but has ", other.dim(), " dimensions instead");
 
-  // NumPy works for 1-dimensional 'other' or batch of 1-dimensional tensors, we need to unsqueeze it
-  // because 2-dimensional tensors are expected in the implementation
-  auto expected_batched_rhs_shape = IntArrayRef(input.sizes().data(), input.dim()-1);  // A.shape[:-1]
-  bool is_rhs_broadcasted = other.dim() == 1 || (input.dim()-1 == other.dim() && other.sizes().equals(expected_batched_rhs_shape));
-  Tensor other_ = is_rhs_broadcasted ? other.unsqueeze(-1) : other;
+  // Two types of 'other' tensors are supported:
+  // - 1-dimensional (1D) tensor or batch of 1D tensors (vector case)
+  // - 2-dimensional (2D) tensor or batch of 2D tensors (matrix case)
+  // original torch.solve supported only the matrix case, while NumPy works for both cases
+  // for the batched input we need to be able to distinguish them
+  auto expected_batched_rhs_shape = IntArrayRef(input.sizes().data(), input.dim()-1);  // input.shape[:-1]
+  bool vector_case = other.dim() == 1 || (input.dim()-1 == other.dim() && other.sizes().equals(expected_batched_rhs_shape));
+
+  bool is_batched_column_major = false;
+  if (vector_case) {
+    is_batched_column_major = result.is_contiguous();
+  } else if (!vector_case && result.dim() >= 2) {
+    is_batched_column_major = result.transpose(-2, -1).is_contiguous();
+  }
+
+  // if 'other' is a batch of 2D tensors, then 'input' can be non-batched and will be broadcasted
+  auto expected_shape = expected_batched_rhs_shape;
+  if (!vector_case && other.dim() > 2) {
+    expected_shape = other.sizes();
+  }
+
+  bool result_equal_expected_shape = result.sizes().equals(expected_shape);
+  bool result_input_same_type = (result.scalar_type() == input.scalar_type());
+
+  // if result is not empty and not in batched column major format
+  bool copy_needed = (result.numel() != 0 && !is_batched_column_major);
+  copy_needed |= !result_input_same_type;  // or result does not have the same dtype as input
+  copy_needed |= (result.numel() != 0 && !result_equal_expected_shape); // or result does not have the expected shape
+  // we have to allocate a temporary tensor
+  if (copy_needed) {
+    Tensor result_tmp = at::empty({0}, input.options());
+    result_tmp = linalg_solve_out_info(result_tmp, infos, input, other);
```

Comment:
Same comment here as on linalg.inv PR about avoiding this recursive call (if possible and if it would improve readability).

Reply:
It was decided in the linalg.inv PR that it is fine to keep this recursive call here.
https://github.com/pytorch/pytorch/pull/51977#discussion_r575097797